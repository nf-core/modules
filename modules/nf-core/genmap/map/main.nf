process GENMAP_MAP {
    tag "$meta.id"
    label 'process_high'

    conda "bioconda::genmap=1.3.0"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/genmap:1.3.0--h1b792b2_1' :
        'biocontainers/genmap:1.3.0--h1b792b2_1' }"

    input:
    tuple val(meta), path(index)
    tuple val(meta2), path(regions)

    output:
    tuple val(meta), path("*.wig")      , optional:true, emit: wig
    tuple val(meta), path("*.bedgraph") , optional:true, emit: bedgraph
    tuple val(meta), path("*.txt")      , optional:true, emit: txt
    tuple val(meta), path("*.csv")      , optional:true, emit: csv
    path "versions.yml"                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "$meta.id"
    def bed = regions ? "--selection ${regions}" : ""

    if ("$index" == "${prefix}") error "Input and output names are the same, set prefix in module configuration to disambiguate!"
    """
    genmap \\
        map \\
        ${args} \\
        ${bed} \\
        --threads ${task.cpus} \\
        --index ${index} \\
        --output ${prefix}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        genmap: \$(genmap --version 2>&1 | sed 's/GenMap version: //; s/SeqAn.*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "$meta.id"
    def token_args = args.tokenize(" ")
    def wig = token_args.contains("-w") || token_args.contains("--wig") ?       "touch ${prefix}.wig"       : ""
    def bg =  token_args.contains("-bg") || token_args.contains("--bedgraph") ? "touch ${prefix}.bedgraph"  : ""
    def txt = token_args.contains("-t") || token_args.contains("--txt") ?       "touch ${prefix}.txt"       : ""
    def csv = token_args.contains("-d") || token_args.contains("--csv") ?       "touch ${prefix}.csv"       : ""

    if ("$index" == "${prefix}") error "Input and output names are the same, set prefix in module configuration to disambiguate!"
    """
    ${wig}
    ${bg}
    ${txt}
    ${csv}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        genmap: \$(genmap --version 2>&1 | sed 's/GenMap version: //; s/SeqAn.*\$//')
    END_VERSIONS
    """
}
