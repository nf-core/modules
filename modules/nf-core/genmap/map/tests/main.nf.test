nextflow_process {

    name "Test Process GENMAP_MAP"
    config "./nextflow.config"
    script "../main.nf"
    process "GENMAP_MAP"

    tag "modules"
    tag "modules_nfcore"
    tag "genmap"
    tag "genmap/index"
    tag "genmap/map"

    setup {
        run("GENMAP_INDEX") {
            script "../../index/main.nf"
            process {
                """
                input[0] = [
                    [id:"test"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists:true)
                ]
                """
            }
        }
    }

    test("sarscov2 - index bed") {

        when {
            params {
                module_args   = "-K 50 -E 2 -w -t -bg -d"
            }
            process {
                """
                input[0] = GENMAP_INDEX.out.index.view()
                input[1] = [
                    [id:"bed"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed', checkIfExists:true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }


    test("sarscov2 - index bed -- stub") {
        options '-stub'
        when {
            params {
                module_args   = "-K 50 -E 2 -w -t -bg -d"
            }
            process {
                """
                input[0] = GENMAP_INDEX.out.index
                input[1] = [
                    [id:"bed"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed', checkIfExists:true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out,
                    path(process.out.versions[0]).yaml
                ).match() }
            )
        }
    }
}