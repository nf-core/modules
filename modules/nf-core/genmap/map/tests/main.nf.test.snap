{
    "sarscov2 - index bed": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.wig:md5,c940a5c35af6ac1c11ecc87b7662fc2a"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.bedgraph:md5,27ca54c920f3ff0c5985384e557a40f2"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.txt:md5,7f9aeaa3d174bb5e6f552efd4f857b4e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.csv:md5,4167a6c7ffbecc1a239c5c1e16b2cbc5"
                    ]
                ],
                "4": [
                    "versions.yml:md5,c510ddb0e56c8961d4cf9d538a789087"
                ],
                "bedgraph": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.bedgraph:md5,27ca54c920f3ff0c5985384e557a40f2"
                    ]
                ],
                "csv": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.csv:md5,4167a6c7ffbecc1a239c5c1e16b2cbc5"
                    ]
                ],
                "txt": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.txt:md5,7f9aeaa3d174bb5e6f552efd4f857b4e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c510ddb0e56c8961d4cf9d538a789087"
                ],
                "wig": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.wig:md5,c940a5c35af6ac1c11ecc87b7662fc2a"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-04-11T16:10:45.401177409"
    },
    "sarscov2 - index bed -- stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.wig:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.bedgraph:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,c510ddb0e56c8961d4cf9d538a789087"
                ],
                "bedgraph": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.bedgraph:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "csv": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "txt": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c510ddb0e56c8961d4cf9d538a789087"
                ],
                "wig": [
                    [
                        {
                            "id": "test"
                        },
                        "test_map.wig:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ]
            },
            {
                "GENMAP_MAP": {
                    "genmap": "1.3.0"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-04-11T16:10:51.374175856"
    }
}